### PR TITLE
GAWB 2778 - Delete error clusters

### PR DESCRIPTION
**Pull request by vkumra-broad**
Changes to allow deletion of error clusters(ClustErrs)

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by vkumra-broad** [comment]
jenkins retest

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala [hunk 1]
```diff
@@ -397,6 +397,7 @@ class GoogleDataprocDAO(protected val dataprocConfig: DataprocConfig,
     executeGoogleRequestAsync(googleProject, clusterName.toString, request).recover {
       // treat a 404 error as a successful deletion
       case CallToGoogleApiFailedException(_, _, 404, _) => ()
+      case CallToGoogleApiFailedException(_, _, _, msg) if msg.contains("it has other pending delete operations against it") => ()
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: can you match against the status code Google returns instead / as well?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala [hunk 2]
```diff
@@ -20,6 +21,94 @@ case class GcsPath(bucketName: GcsBucketName, relativePath: GcsRelativePath) {
   def toUri: String = s"$GCS_SCHEME://${bucketName.name}/${relativePath.name}"
 }
 
+object GcsPath {
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala]: Do we need to copy-paste all of this? Can't we just use workbench-libs instead?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala [hunk 3]
```diff
@@ -128,7 +129,20 @@ class LeonardoSpec extends FreeSpec with Matchers with Eventually with ParallelT
...
+      val cluster = createAndMonitor(googleProject, name, request, ClusterStatus.Running)
+      testCode(cluster)
+    }
+
+    // delete before checking testCode status, which may throw
+    deleteAndMonitor(googleProject, name)
+    testResult.get
+  }
+
+  def withNewErroredCluster[T](googleProject: GoogleProject)(testCode: Cluster => T): T = {
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala]: Thank you, I need this

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 4]
```diff
@@ -74,7 +74,7 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
       case None =>
         val augmentedClusterRequest = addClusterDefaultLabels(serviceAccount, googleProject, clusterName, clusterRequest)
         val createFuture = for {
-          // Create the cluster in Google
+        // Create the cluster in Google
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: The indentation seemed correct before

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 5]
```diff
@@ -141,7 +141,7 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
   private[service] def createGoogleCluster(serviceAccount: WorkbenchEmail, googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[(Cluster, GcsBucketName, Option[ServiceAccountKey])] = {
     val initBucketName = generateUniqueBucketName(clusterName.string)
     for {
-      // Validate that the Jupyter extension URI is a valid URI and references a real GCS object
+    // Validate that the Jupyter extension URI is a valid URI and references a real GCS object
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Indentation is off

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 6]
```diff
@@ -195,8 +195,8 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
   /* Create a google bucket and populate it with init files */
   private[service] def initializeBucket(googleProject: GoogleProject, clusterName: ClusterName, bucketName: GcsBucketName, clusterRequest: ClusterRequest, serviceAccountEmail: WorkbenchEmail, serviceAccountKey: Option[ServiceAccountKey]): Future[GcsBucketName] = {
     for {
-      // Note the bucket is created in Leo's project, not the cluster's project.
-      // ACLs are granted so the cluster's service account can access the bucket at initialization time.
+    // Note the bucket is created in Leo's project, not the cluster's project.
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: ditto

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 7]
```diff
@@ -236,7 +236,7 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
     val googleSignInJsContent = templateResource(clusterResourcesConfig.jupyterGoogleSignInJs, replacements)
 
     for {
-      // Upload the init script to the bucket
+    // Upload the init script to the bucket
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: ditto

**Comment by rtitle** [comment]
I didn't test, but it looks good, pending minor comments.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala (hunk 4 above)
Comment (vkumra-broad) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Intellij auto indent has failed me

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala (hunk 2 above)
Comment (helgridly) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala]: +1

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala [hunk 8]
```diff
@@ -66,7 +155,7 @@ case class Cluster(clusterName: ClusterName,
 
 case class ClusterRequest(bucketPath: GcsBucketName,
                           labels: LabelMap,
-                          jupyterExtensionUri: Option[GcsPath] = None)
+                          jupyterExtensionUri: Option[String] = None)
```
Comment (helgridly) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala]: Why this change? This allows us to immediately barf at the user when interpreting the JSON if someone gives us something that's not shaped like a bucket path.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala [hunk 9]
```diff
@@ -38,6 +38,7 @@ class LeonardoSpec extends FreeSpec with Matchers with Eventually with ParallelT
   val project = GoogleProject(LeonardoConfig.Projects.default)
   val sa = GoogleServiceAccount(LeonardoConfig.Leonardo.notebooksServiceAccountEmail)
   val bucket = GcsBucketName("mah-bukkit")
+  val incorrectJupyterExtensionUri = "gs://leonardo-swat-test-bucket-do-not-delete"
```
Comment (helgridly) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala]: This is just a bucket. Doesn't it need to be a nonexistent file inside this bucket?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala [hunk 10]
```diff
@@ -177,6 +191,12 @@ class LeonardoSpec extends FreeSpec with Matchers with Eventually with ParallelT
       }
     }
 
+    "should error on cluster create and delete the cluster" in {
+      withNewErroredCluster(project) { _ =>
+        // no-op; just verify that it launches
```
Comment (helgridly) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala]: confused. shouldn't there be assertions in here?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala (hunk 9 above)
Comment (vkumra-broad) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala]: Passing an empty bucket is causes cluster creation to error, I haven't tried it with a non existent file. Is there a reason to use a non existent file instead?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala (hunk 10 above)
Comment (vkumra-broad) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala]: The assertions are inside of the withNewErroredCluster

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala (hunk 8 above)
Comment (vkumra-broad) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala]: I was running into issues with deserializing the GcsPath, for the tests  its easier to send this request parameter as a String instead of a GcsPath

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala (hunk 9 above)
Comment (helgridly) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala]: A file and a bucket aren't the same thing and the error is almost certainly different. Also, isn't that the bucket we made yesterday? I thought we made that public? Or did you set it back to private?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala (hunk 8 above)
Comment (helgridly) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala]: One day we'll publish leo-model and we'll be straight back to this issue, though.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala (hunk 10 above)
Comment (helgridly) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala]: Ah ok, thanks.

File: automation/project/Dependencies.scala [hunk 11]
```diff
@@ -4,6 +4,8 @@ object Dependencies {
   val jacksonV = "2.8.4"
   val akkaV = "2.4.17"
   val akkaHttpV = "10.0.5"
+  val workbenchGoogleV  = "0.10-d97f551"
```
Comment (rtitle) [inline automation/project/Dependencies.scala]: Oops it looks like we both did this:
https://github.com/broadinstitute/leonardo/pull/80

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala [hunk 12]
```diff
@@ -1,12 +1,14 @@
 package org.broadinstitute.dsde.workbench.leonardo
 
-import java.net.URL
+import java.net.{URI, URL}
 import java.time.Instant
 import java.util.UUID
-
+import com.google.common.net.UrlEscapers
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala]: Probably unused import now?

**Comment by vkumra-broad** [comment]
jenkins retest

**Comment by vkumra-broad** [comment]
jenkins retest

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala (hunk 8 above)
Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala]: Current Leo Swat has lots of this.  The only reason this wasn't covered before is because I only had `None` for these in my tests.

Comment (helgridly) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala]: fair. we'll have to fix this At Some Point but we can wait until then

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala [hunk 13]
```diff
@@ -53,7 +54,7 @@ class LeonardoSpec extends FreeSpec with Matchers with Eventually with ParallelT
     val dummyPetSa = WorkbenchUserServiceAccountEmail("dummy")
     val expected = requestedLabels ++ DefaultLabels(clusterName, googleProject, gcsBucketName, dummyPetSa, None).toMap - "serviceAccount"
 
-    (seen - "serviceAccount") shouldBe (expected - "serviceAccount")
+    (seen - "serviceAccount" - "notebookExtension") shouldBe (expected - "serviceAccount" - "notebookExtension")
```
Comment (jmthibault79) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala]: Why is this removed from the comparison?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala [hunk 14]
```diff
@@ -82,13 +83,13 @@ class LeonardoSpec extends FreeSpec with Matchers with Eventually with ParallelT
     clusterCheck(Leonardo.cluster.get(googleProject, clusterName), clusterRequest.labels, clusterName, Seq(ClusterStatus.Creating))
 
     // wait for "Running" or error (fail fast)
-    val runningCluster = eventually {
+    val actualCluster = eventually {
       clusterCheck(Leonardo.cluster.get(googleProject, clusterName), clusterRequest.labels, clusterName, Seq(ClusterStatus.Running, ClusterStatus.Error))
     } (clusterPatience)
 
     // now check that it didn't timeout or error
-    runningCluster.status shouldBe ClusterStatus.Running
```
Comment (jmthibault79) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala]: I think it's more clear to move this check outside `createAndMonitor`

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala [hunk 15]
```diff
@@ -128,7 +129,20 @@ class LeonardoSpec extends FreeSpec with Matchers with Eventually with ParallelT
...
+      testCode(cluster)
+    }
+
+    // delete before checking testCode status, which may throw
+    deleteAndMonitor(googleProject, name)
+    testResult.get
+  }
+
+  def withNewErroredCluster[T](googleProject: GoogleProject)(testCode: Cluster => T): T = {
+    val name = ClusterName(s"automation-test-a${Random.alphanumeric.take(10).mkString.toLowerCase}z")
```
Comment (jmthibault79) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala]: These can use `makeRandomId()`

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala (hunk 10 above)
Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala]: Heh Rob didn't like it when I started this pattern either.  :)

Should change the text to "verify that it errors" or similar

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala [hunk 16]
```diff
@@ -82,23 +84,23 @@ object ClusterStatus extends Enumeration {
...
-      )
+      clusterName = clusterName,
+      googleId = googleId,
+      googleProject = googleProject,
+      googleServiceAccount = serviceAccount,
+      googleBucket = clusterRequest.bucketPath,
+      machineConfig = MachineConfig(clusterRequest.machineConfig, clusterDefaultsConfig),
+      clusterUrl = getClusterUrl(googleProject, clusterName),
+      operationName = operationName,
+      status = clusterStatus,
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: Hm.  I don't like this from a semantic standpoint.  I want `create()` to always produce a `Creating` cluster.  It's easy to change the status in tests post-creation

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala [hunk 17]
```diff
@@ -252,6 +252,26 @@ class LeonardoServiceSpec extends TestKit(ActorSystem("leonardotest")) with Flat
     gdDAO.clusters should not contain key (clusterName)
   }
 
+  it should "delete a cluster that has status Error" in isolatedDbTest {
+    // check that the cluster does not exist
+    gdDAO.clusters should not contain key (clusterName)
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala]: This is not the same cluster as the rest of the test
